doc/code_style_guidelines.md

```diff
@@ -0,0 +1,218 @@
+* Try to minimize the use of exceptions.
+* Avoid operator overloading like the plague.
+* Avoid templates unless they are very convenient.
+* Usage of macros is not discouraged, just don't overdo it [like X.org](http://cgit.freedesktop.org/xorg/xserver/tree/randr/rrscreen.c?id=01e18af17f8dc91451fbd0902049045afd1cea7e#n325). It's better to use inline functions or lambdas instead.
+
+
+# Classes
+
+* **Class names are *PascalCase*, method names are *camelCase*.**
+* Don't put actual code in header files, unless it's a 4-liner, an inline function, or part of a template.
+* Class definitions should go in header files.
+* Substantial methods (over 4 lines) should be defined outside of the class definition.
+* Functions not part of any class should use `lowercase_underscore_style()`.
+
+
+# Comments
+
+* Doxygen comments are acceptable, but **please** put them in the header file.
+* Don't make uninformative comments like this:
+
+```cpp
+// Draw "Loading" screen
+draw_load_screen(L"Loading...", driver, font);
+```
+
+* Add comments to explain a non-trivial but important detail about the code, or explain behavior that is not obvious.
+* For comments with text, be sure to add a space between the text and the comment tokens:
+
+```cpp
+DoThingHere();  // This does thing    <--- yes!
+DoThingHere();  /* This does thing */ <--- yes!
+
+DoThingHere();  //This does thing      <--- no!
+DoThingHere();  /*This does thing*/    <--- no!
+DoThingHere();//This does thing        <--- no!
+```
+
+
+# Use STL, avoid Irrlicht containers, and no, Boost will not even be considered, so forget it
+
+* In general, adding new dependencies is considered serious business.
+* We are using C++14; Boost will never be an option.
+
+
+# Don't let things get too large
+
+* **Try to keep lines under 95 characters.** It's okay if it goes over by a few, but do not exaggerate. (Note that this column count assumes 4-space indents.)
+* Functions should not have over 200 lines of code – if you are concerned with having to pass too many parameters to child functions, make whatever it is into a class.
+* Don't let files get too large (over 1500 lines of code). Currently, existing huge files (`game.cpp`, `server.cpp`, …) are in the slow process of being cleaned up.
+
+
+# Files
+
+* Files should be named using *snake_case* style.
+* Files should have includes for everything that they depend on. Don't depend on, eg, `"util/numeric.h"` including `<string>`!
+* Uniqueness when compiling headers is ensured by using `#pragma once`. ([Accepted by all coredevs](https://github.com/minetest/minetest/issues/6259))
+
+```cpp
+#pragma once
+
+#include <string>
+
+class Foo {
+};
+
+```
+
+* All files should include the appropriate license header.
+
+
+# Miscellaneous
+
+* <span style="color: red">Do **not** use `or`, use `||`.</span>
+* Set pointer values to `nullptr` (C++11), not 0.
+* When using floats, add the `f` suffix, e.g. `float k = 0.0f;` and not `float k = 0.0;`.
```

Comment:
Exception: Integers should work fine (`0` in particular might be a bad example); only doubles might be problematic?

Reply:
Would be content change.
(And no, we want `0.0f` code-style wise. No reason for inconsistency for `0`.)